src/Compiler/Microsoft.NET.Sdk.Razor.SourceGenerators/SourceGeneratorProjectEngine.cs

```diff
@@ -0,0 +1,119 @@
+    {
+        for (int i = 0; i < Engine.Phases.Count; i++)
+        {
+            if (Engine.Phases[i] is DefaultRazorTagHelperContextDiscoveryPhase)
+            {
+                discoveryPhaseIndex = i;
+            }
+            else if (Engine.Phases[i] is DefaultRazorTagHelperRewritePhase)
+            {
+                rewritePhaseIndex = i;
+            }
+            else if (discoveryPhaseIndex >= 0 && rewritePhaseIndex >= 0)
+            {
+                break;
+            }
+        }
+        Debug.Assert(discoveryPhaseIndex >= 0);
+        Debug.Assert(rewritePhaseIndex >= 0);
+    }
+
+    public SourceGeneratorRazorCodeDocument ProcessInitialParse(RazorProjectItem projectItem, bool designTime)
+    {
+        var codeDocument = designTime 
+                            ? CreateCodeDocumentDesignTimeCore(projectItem)
+                            : CreateCodeDocumentCore(projectItem);
+
+
+        ProcessPartial(codeDocument, 0, discoveryPhaseIndex);
+
+        // record the syntax tree, before the tag helper re-writing occurs
+        codeDocument.SetPreTagHelperSyntaxTree(codeDocument.GetSyntaxTree());
+        return new SourceGeneratorRazorCodeDocument(codeDocument);
+    }
+
+    public SourceGeneratorRazorCodeDocument ProcessTagHelpers(SourceGeneratorRazorCodeDocument sgDocument, IReadOnlyList<TagHelperDescriptor> tagHelpers, bool checkForIdempotency)
+    {
+        Debug.Assert(sgDocument.CodeDocument.GetPreTagHelperSyntaxTree() is not null);
+
+        int startIndex = discoveryPhaseIndex;
+        var codeDocument = sgDocument.CodeDocument;
+        var previousTagHelpers = codeDocument.GetTagHelpers();
+        if (checkForIdempotency && previousTagHelpers is not null)
+        {
+            // compare the tag helpers with the ones the document last used
+            if (Enumerable.SequenceEqual(tagHelpers, previousTagHelpers))
+            {
+                // tag helpers are the same, nothing to do!
+                return sgDocument;
+            }
+            else
+            {
+                // tag helpers have changed, figure out if we need to re-write
+                var oldContextHelpers = codeDocument.GetTagHelperContext().TagHelpers;
+
+                // re-run the scope check to figure out which tag helpers this document can see
+                codeDocument.SetTagHelpers(tagHelpers);
+                Engine.Phases[discoveryPhaseIndex].Execute(codeDocument);
+
+                // Check if any new tag helpers were added or ones we previously used were removed
+                var newContextHelpers = codeDocument.GetTagHelperContext().TagHelpers;
+                var added = newContextHelpers.Except(oldContextHelpers);  //PROTOTYPE: can we decide if a tag helper is 'modified' rather than being added? Basically if it still has the same name, then surely we know it still isn't used?
+                var referencedByRemoved = codeDocument.GetReferencedTagHelpers().Except(newContextHelpers);
```

Comment:
Consider adding comments or improving naming. It took me some time to figure these sets mean the following:
1. previously visible tag helpers (`oldContextHelpers`)
2. currently visible tag helpers (`newContextHelpers`)
3. newly visible tag helpers (`added`)
4. previously used tag helpers (`codeDocument.GetReferencedTagHelpers()`)
5. removed previously used (newly unused) tag helpers (`referencedByRemoved`)